File: src-tauri/src/services/speech.rs
use crate::logging;
use base64::Engine;

/// Transcribe audio using the Mistral Voxtral API.
/// Accepts base64-encoded audio data and a MIME type (e.g. "audio/webm").
pub async fn transcribe(api_key: &str, audio_base64: &str, media_type: &str) -> Result<String, String> {
    logging::info(&format!("speech::transcribe: media_type={}, audio_len={}", media_type, audio_base64.len()));

    let audio_bytes = base64::engine::general_purpose::STANDARD
        .decode(audio_base64)
        .map_err(|e| format!("Failed to decode audio: {}", e))?;

    logging::info(&format!("speech::transcribe: decoded {} bytes", audio_bytes.len()));

    let ext = match media_type {
        "audio/webm" => "webm",
        "audio/ogg" => "ogg",
        "audio/mp4" => "mp4",
        "audio/wav" => "wav",
        "audio/mpeg" => "mp3",
        _ => "webm",
    };

    let file_part = reqwest::multipart::Part::bytes(audio_bytes)
        .file_name(format!("recording.{}", ext))
        .mime_str(media_type)
        .map_err(|e| format!("Failed to create multipart: {}", e))?;

    let form = reqwest::multipart::Form::new()
        .text("model", "voxtral-mini-latest")
        .part("file", file_part);

    let client = reqwest::Client::new();
    let resp = client
        .post("https://api.mistral.ai/v1/audio/transcriptions")
        .header("Authorization", format!("Bearer {}", api_key))
        .multipart(form)
        .send()
        .await
        .map_err(|e| format!("Voxtral API request failed: {}", e))?;

    if !resp.status().is_success() {
        let status = resp.status();
        let text = resp.text().await.unwrap_or_default();
        let msg = format!("Voxtral API error ({}): {}", status, text);
        logging::error(&msg);
        return Err(msg);
    }

    #[derive(serde::Deserialize)]
    struct VoxtralResponse {
        text: String,
    }

    let data: VoxtralResponse = resp.json().await
        .map_err(|e| format!("Failed to parse Voxtral response: {}", e))?;

    logging::info(&format!("speech::transcribe: result='{}' ({} chars)", &data.text[..data.text.len().min(80)], data.text.len()));
    Ok(data.text)
}

/// Transcribe a WAV file from disk.
pub async fn transcribe_file(api_key: &str, file_path: &str) -> Result<String, String> {
    logging::info(&format!("speech::transcribe_file: path={}", file_path));

    let audio_bytes = std::fs::read(file_path)
        .map_err(|e| format!("Failed to read audio file: {}", e))?;

    logging::info(&format!("speech::transcribe_file: read {} bytes", audio_bytes.len()));

    if audio_bytes.is_empty() {
        return Err("Recording file is empty — no audio was captured.".to_string());
    }

    let file_part = reqwest::multipart::Part::bytes(audio_bytes)
        .file_name("recording.wav")
        .mime_str("audio/wav")
        .map_err(|e| format!("Failed to create multipart: {}", e))?;

    let form = reqwest::multipart::Form::new()
        .text("model", "voxtral-mini-latest")
        .part("file", file_part);

    let client = reqwest::Client::new();
    let resp = client
        .post("https://api.mistral.ai/v1/audio/transcriptions")
        .header("Authorization", format!("Bearer {}", api_key))
        .multipart(form)
        .send()
        .await
        .map_err(|e| format!("Voxtral API request failed: {}", e))?;

    if !resp.status().is_success() {
        let status = resp.status();
        let text = resp.text().await.unwrap_or_default();
        let msg = format!("Voxtral API error ({}): {}", status, text);
        logging::error(&msg);
        return Err(msg);
    }

    #[derive(serde::Deserialize)]
    struct VoxtralResponse {
        text: String,
    }

    let data: VoxtralResponse = resp.json().await
        .map_err(|e| format!("Failed to parse Voxtral response: {}", e))?;

    logging::info(&format!("speech::transcribe_file: result='{}' ({} chars)", &data.text[..data.text.len().min(80)], data.text.len()));
    Ok(data.text)
}

/// Text-to-speech using the Mistral Voxtral TTS API. Returns base64-encoded audio.
pub async fn text_to_speech(api_key: &str, text: &str) -> Result<String, String> {
    logging::info(&format!("speech::tts: {} chars", text.len()));

    let body = serde_json::json!({
        "model": "voxtral-mini-tts-2603",
        "input": text,
        "voice": "en_paul_neutral",
        "response_format": "wav",
    });

    let client = reqwest::Client::new();
    let resp = client
        .post("https://api.mistral.ai/v1/audio/speech")
        .header("Authorization", format!("Bearer {}", api_key))
        .header("Content-Type", "application/json")
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("Voxtral TTS request failed: {}", e))?;

    if !resp.status().is_success() {
        let status = resp.status();
        let text = resp.text().await.unwrap_or_default();
        let msg = format!("Voxtral TTS error ({}): {}", status, text);
        logging::error(&msg);
        return Err(msg);
    }

    #[derive(serde::Deserialize)]
    struct TtsResponse {
        audio_data: String,
    }

    let data: TtsResponse = resp.json().await
        .map_err(|e| format!("Failed to parse Voxtral TTS response: {}", e))?;

    logging::info(&format!("speech::tts: got {} bytes base64 audio", data.audio_data.len()));
    Ok(data.audio_data)
}

File: web/src/speech.ts
// Mistral Voxtral (speech) and vision calls, ported from the desktop app's
// `services::speech` / `services::vision`. The browser can't hold the API key,
// so these stay server-side and read it from the saved agent config.

import { getAgentConfig } from "./config.js";

const TRANSCRIBE_MODEL = "voxtral-mini-latest";
const TTS_MODEL = "voxtral-mini-tts-2603";
const VISION_MODEL = "mistral-large-2512";

function requireKey(): string {
  const key = getAgentConfig().speech_api_key;
  if (!key) {
    throw new Error("No speech API key configured. Add your Mistral API key under Speech in Settings.");
  }
  return key;
}

const EXTENSIONS: Record<string, string> = {
  "audio/webm": "webm",
  "audio/ogg": "ogg",
  "audio/mp4": "mp4",
  "audio/wav": "wav",
  "audio/mpeg": "mp3",
};

async function mistral(path: string, init: RequestInit, key: string): Promise<Response> {
  const res = await fetch(`https://api.mistral.ai/v1${path}`, {
    ...init,
    headers: { ...(init.headers as Record<string, string>), Authorization: `Bearer ${key}` },
  });
  if (!res.ok) {
    throw new Error(`Mistral API error (${res.status}): ${await res.text()}`);
  }
  return res;
}

/** Transcribe base64 audio captured in the browser. */
export async function transcribe(audioBase64: string, mediaType: string): Promise<string> {
  const key = requireKey();
  const bytes = Buffer.from(audioBase64, "base64");
  if (bytes.length === 0) throw new Error("Recording is empty — no audio was captured.");

  const form = new FormData();
  form.append("model", TRANSCRIBE_MODEL);
  form.append("file", new Blob([bytes], { type: mediaType }), `recording.${EXTENSIONS[mediaType] ?? "webm"}`);

  const res = await mistral("/audio/transcriptions", { method: "POST", body: form }, key);
  return ((await res.json()) as { text: string }).text;
}

/** Text-to-speech. Returns base64 WAV, matching the desktop command. */
export async function textToSpeech(text: string): Promise<string> {
  const key = requireKey();
  const res = await mistral(
    "/audio/speech",
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ model: TTS_MODEL, input: text, voice: "en_paul_neutral", response_format: "wav" }),
    },
    key,
  );
  return ((await res.json()) as { audio_data: string }).audio_data;
}

/** Extract the information in an image that the agent needs to act on it. */
export async function describeImage(imageBase64: string, mediaType: string, userPrompt: string): Promise<string> {
  const key = requireKey();

  const extractionPrompt = userPrompt.trim()
    ? `The user has shared this image with their AI assistant and asks: "${userPrompt.replace(/"/g, "'")}"\n\n` +
      "Extract the relevant information from the image so the assistant can fulfill the request. " +
      "If the image contains a list, transcribe each item exactly. If it contains text, transcribe it. " +
      "If it contains a diagram or structure, describe the structure. " +
      "Return ONLY the extracted information — no preamble, no commentary on what the user should do."
    : "Describe this image in detail. Extract any text, lists, diagrams, or structured information " +
      "that would be useful for further processing. Be thorough but concise.";

  const res = await mistral(
    "/chat/completions",
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        model: VISION_MODEL,
        messages: [
          {
            role: "user",
            content: [
              { type: "text", text: extractionPrompt },
              { type: "image_url", image_url: `data:${mediaType};base64,${imageBase64}` },
            ],
          },
        ],
      }),
    },
    key,
  );

  const data = (await res.json()) as { choices?: { message: { content: string } }[] };
  const text = data.choices?.[0]?.message.content;
  if (text === undefined) throw new Error("Mistral vision returned no choices");
  return text;
}
